algorithms/src/snark/marlin/data_structures/proof.rs

```diff
@@ -233,35 +273,45 @@ pub struct Proof<E: PairingEngine> {
 impl<E: PairingEngine> Proof<E> {
     /// Construct a new proof.
     pub fn new(
-        batch_size: usize,
+        batch_sizes: BTreeMap<CircuitId, usize>,
         commitments: Commitments<E>,
         evaluations: Evaluations<E::Fr>,
         msg: ahp::prover::ThirdMessage<E::Fr>,
         pc_proof: sonic_pc::BatchLCProof<E>,
     ) -> Result<Self, SNARKError> {
-        if commitments.witness_commitments.len() != batch_size {
-            return Err(SNARKError::BatchSizeMismatch);
+        let mut total_instances = 0;
+        let batch_sizes: Vec<usize> = batch_sizes.into_values().collect();
+        for (z_b_evals, batch_size) in evaluations.z_b_evals.iter().zip(&batch_sizes) {
+            total_instances += batch_size;
+            if z_b_evals.len() != *batch_size {
+                return Err(SNARKError::BatchSizeMismatch);
+            }
         }
-        if evaluations.z_b_evals.len() != batch_size {
+        if commitments.witness_commitments.len() != total_instances {
             return Err(SNARKError::BatchSizeMismatch);
         }
-        Ok(Self { batch_size, commitments, evaluations, msg, pc_proof })
+        Ok(Self { batch_sizes, commitments, evaluations, msg, pc_proof })
     }
 
-    pub fn batch_size(&self) -> Result<usize, SNARKError> {
-        if self.commitments.witness_commitments.len() != self.batch_size {
-            return Err(SNARKError::BatchSizeMismatch);
+    pub fn batch_sizes(&self) -> Result<&[usize], SNARKError> {
+        let mut total_instances = 0;
+        for (z_b_evals_i, &batch_size) in self.evaluations.z_b_evals.iter().zip(self.batch_sizes.iter()) {
+            total_instances += batch_size;
+            if z_b_evals_i.len() != batch_size {
+                return Err(SNARKError::BatchSizeMismatch);
+            }
         }
-        if self.evaluations.z_b_evals.len() != self.batch_size {
+        if self.commitments.witness_commitments.len() != total_instances {
             return Err(SNARKError::BatchSizeMismatch);
         }
-        Ok(self.batch_size)
+        Ok(&self.batch_sizes)
     }
 }
 
 impl<E: PairingEngine> CanonicalSerialize for Proof<E> {
     fn serialize_with_mode<W: Write>(&self, mut writer: W, compress: Compress) -> Result<(), SerializationError> {
-        CanonicalSerialize::serialize_with_mode(&self.batch_size, &mut writer, compress)?;
+        let batch_sizes: Vec<u64> = self.batch_sizes.iter().map(|x| *x as u64).collect();
```

Comment:
```suggestion
        let batch_sizes: Vec<u64> = self.batch_sizes.iter().map(|x| u64::try_from(*x)).collect::<Result<_, _>>()?;
```
there's been some discussions regarding possible use on machines with 128bit architecture, so casting from `usize` to `u64` might not always work

Reply:
also, an alternative: `self.batch_sizes.iter().copied().map(|u64::try_from).collect::<Result<_, _>>()?;`

Reply:
2d3d0c701